src/chocolatey/infrastructure.app/domain/ShimRegistry.cs

```diff
@@ -0,0 +1,308 @@
+            var exeFiles = get_shim_exe_files();
+
+            foreach (string file in _store.get_files_to_update(exeFiles))
+            {
+                _store.add_record(get_shim_data(file));
+            }
+        }
+
+        /// <summary>
+        /// Gets the exe files in the shim directory.
+        /// </summary>
+        /// <returns>The exe files.</returns>
+        private IEnumerable<string> get_shim_exe_files()
+        {
+            return _fileSystem.get_files(_shimsLocation, pattern: "*.exe");
+        }
+
+        /// <summary>
+        /// Extracts any shim data from an exe file.
+        /// </summary>
+        /// <param name="exeFile">The exe file</param>
+        /// <returns>A record containing full or partial shim data.</returns>
+        private ShimRecord get_shim_data(string exeFile)
+        {
+            var item = new ShimRecord(exeFile);
+
+            // check the exe is shimgen generated from the file info
+            FileVersionInfo info = FileVersionInfo.GetVersionInfo(exeFile);
+
+            // note that this will not catch chocolatey specific shims
+            if (!info.ProductName.contains("shimgen")) return item;
+
+            // extract the target file from the binary
+            string target = find_shim_target(exeFile);
+
+            item.TargetFile = check_and_get_targetFile(target);
+
+            // try and extract the package name
+            if (item.TargetFile.contains(_packagesLocation))
+            {
+                var match = _packageRegex.Match(item.TargetFile).Groups[1];
+                if (match != null)
+                {
+                    item.PackageName = match.Value;
+                }
+            }
+
+            return item;
+        }
+
+        /// <summary>
+        /// Searches an exe file for the shim target file name.
+        /// </summary>
+        /// <param name="exeFile">The exe file.</param>
+        /// <returns>The target file name or null.</returns>
+        private string find_shim_target(string exeFile)
+        {
+            string target = null;
+            byte[] bytes = File.ReadAllBytes(exeFile);
+
+            // search for the pattern: file at '<filename>'
+            int startIndex = 0;
+            string pattern = "file at '";
+            int foundIndex = search_binary(bytes, pattern, startIndex);
+
+            if (foundIndex == -1)
+            {
+                return target;
+            }
+
+            // search for the closing single-quote
+            startIndex = foundIndex + (pattern.Length * 2);
+            pattern = "'";
+            foundIndex = search_binary(bytes, pattern, startIndex);
+
+            if (foundIndex != -1)
+            {
+                int count = foundIndex - startIndex;
+                var slice = new byte[count];
+                Array.Copy(bytes, startIndex, slice, 0, count);
+                target = Encoding.Unicode.GetString(slice);
+            }
+
+            return target;
+        }

```

Comment:
Ha! That's a very interesting way to get at the data! How does it handle with a large number of shims (300+)?

Reply:
Guessing a lot better than running and text parsing for sure!

Reply:
It's fine with a large number of shims (like mingw and other unixy packages), but as I mentioned in  **Concerns 1** it is hacky and brittle. Far better would be for shimgen to write the target file to a field in its FileVersionInfo data in a simple formatted way. The `SpecialBuild` field looks like it would be appropriate location.

We already use this data to check for a shimgen exe anyway:
https://github.com/chocolatey/choco/blob/26ac2df953ea1c8af2c03c33525ac0600c8e46f6/src/chocolatey/infrastructure.app/domain/ShimRegistry.cs#L189-L192

So it would be even faster to grab the target file from `info.SpecialBuild`, for example.